src/coreclr/System.Private.CoreLib/src/Internal/Runtime/InteropServices/InMemoryAssemblyLoader.cs

```diff
@@ -2,48 +2,85 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Diagnostics.CodeAnalysis;
 using System.Runtime.InteropServices;
 using System.Runtime.Loader;
+using System.Runtime.Versioning;
 
 namespace Internal.Runtime.InteropServices
 {
     /// <summary>
     /// This class enables the .NET IJW host to load an in-memory module as a .NET assembly
     /// </summary>
+    [SupportedOSPlatform("windows")]
     public static class InMemoryAssemblyLoader
     {
-#if TARGET_WINDOWS
         private static bool IsSupported { get; } = InitializeIsSupported();
-
         private static bool InitializeIsSupported() => AppContext.TryGetSwitch("System.Runtime.InteropServices.EnableCppCLIHostActivation", out bool isSupported) ? isSupported : true;
-#endif
 
         /// <summary>
-        /// Loads into an isolated AssemblyLoadContext an assembly that has already been loaded into memory by the OS loader as a native module.
+        /// Loads an assembly that has already been loaded into memory by the OS loader as a native module
+        /// into an isolated AssemblyLoadContext.
         /// </summary>
         /// <param name="moduleHandle">The native module handle for the assembly.</param>
         /// <param name="assemblyPath">The path to the assembly (as a pointer to a UTF-16 C string).</param>
+        [RequiresUnreferencedCode("C++/CLI is not trim-compatible", Url = "https://aka.ms/dotnet-illink/nativehost")]
         public static unsafe void LoadInMemoryAssembly(IntPtr moduleHandle, IntPtr assemblyPath)
         {
-#if TARGET_WINDOWS
             if (!IsSupported)
                 throw new NotSupportedException("This API is not enabled in trimmed scenarios. see https://aka.ms/dotnet-illink/nativehost for more details");
 
+            LoadInMemoryAssemblyInContextImpl(moduleHandle, assemblyPath);
+        }
+
+        /// <summary>
+        /// Loads into an assembly that has already been loaded into memory by the OS loader as a native module
+        /// into the specified load context.
+        /// </summary>
+        /// <param name="moduleHandle">The native module handle for the assembly.</param>
+        /// <param name="assemblyPath">The path to the assembly (as a pointer to a UTF-16 C string).</param>
+        /// <param name="loadContext">Load context (currently must be IntPtr.Zero)</param>
+        [UnmanagedCallersOnly]
+        [RequiresUnreferencedCode("C++/CLI is not trim-compatible", Url = "https://aka.ms/dotnet-illink/nativehost")]
+        internal static unsafe void LoadInMemoryAssemblyInContext(IntPtr moduleHandle, IntPtr assemblyPath, IntPtr loadContext)
```

Comment:
We could go either with the same mechanism as with the `LoadInMemoryAssembly` (remove the RUC on this method, suppress it inside).
Or since this is internal API we could root it conditionally based on the same feature switch the `IsSupported` uses - in that case linker would remove this entire method in trimmed apps (better for size).

Reply:
Oh, I guess we changed that based on https://github.com/dotnet/runtime/pull/62903? But just for ComponentActivator and not COM or C++/CLI.

Reply:
So I'd also need to update the sdk so that if you set `EnableCppCLIHostActivation=true`, it also sets `_EnableConsumingManagedCodeFromNativeHosting=true` (or `EnableNativeHostActivation` since it is being renamed) for an app to at least try to go through C++/CLI activation.

And then with the default settings (`EnableCppCLIHostActivation=false`), even with `LoadInMemoryAssemblyInContext` rooted unconditionally as I have it right now, users won't get a nice message. Unless I also root `ComponentActivator.GetFunctionPointer` when `System.Runtime.InteropServices.EnableCppCLIHostActivation=true` and do some sort of handling so that https://github.com/dotnet/runtime/blob/ba962fd91ced147ee7b27f03f7d2a24b288f2e38/src/libraries/System.Private.CoreLib/src/Internal/Runtime/InteropServices/ComponentActivator.cs#L117-L118
also enables support when C++/CLI activation is supported.

The non-nice message is:
```
Unhandled exception. System.Runtime.InteropServices.SEHException (0x80004005): External component has thrown an exception.
```
for trying to call a native entry point (which then calls into managed) in a C++/CLI library.

Reply:
Hmm - that's a good point - I didn't realize that by relying on the `get_function_pointer` it breaks the existing assumptions around feature switches... I told you there are some dragons hidden here 😇 .

Is this because C++/CLI is the only hosting which can propagate full error messages?

I guess what you describe would be a solution - enable native hosting if CPP/CLI is enabled. I would probably do it fully - if you want C++/CLI, you also get full native hosting support - I would not try to do magic in the ComponentActivator. Trimming with C++/CLI activation enabled is a very bad idea anyway.

Reply:
I think COM also would get the full error messaged when it is managed app loading managed COM component (didn't test it though).

I definitely don't want to do any magic in `ComponentActivator` - my thought for potentially having to do something there would purely be for trying to get a nice error message when someone tries to use C++/CLI when it is disabled.

Reply:
I think I finally understand it - if the feature switches are off, we remove the main entry point and that leads to bad error experience. I think it would make sense to keep the `GetFunctionPointer` always. I like it a lot as the main entry point for all native->managed (even though that's not what it does today), and as such we should keep it. On Windows eventually we could migrate all hosts to use that, in which case we could completely remove COM/C++/CLI entry points if the feature switches are off, since we would only need the `GetFunctionPointer` to fail. That would actually be an overall size win. On non-Windows this will be a very small size regression, but I think it's worth it.

Reply:
I definitely agree that `GetFunctionPointer` is nice as the main entry point. In order to have the nicer error experience (like the managed exception with an informative message), we would probably still need it to have some hook to do something for the COM and C++/CLI cases other than just return an error code if the native hosting feature switch is off.

We could also only keep it for Windows, since those are the provided host scenarios we currently have and only do it for non-Windows when we need to - not sure if it is worth the split/discrepancy .

I'll try always keeping `GetFunctionPointer` and working out a decent error experience.